plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/WarPublisher.java

```diff
@@ -101,27 +102,38 @@ private static void publish(IProject project, IPath destination, boolean explode
       for (IModule child : delegate.getChildModules()) {
         ModuleDelegate childDelegate = (ModuleDelegate)
             child.loadAdapter(ModuleDelegate.class, monitor);
-        IWebFragmentModule webFragmentModule = (IWebFragmentModule)
-            child.loadAdapter(IWebFragmentModule.class, monitor);
-        if (childDelegate == null || webFragmentModule == null || !webFragmentModule.isBinary()) {
-          logger.log(Level.WARNING, "child modules other than binary web-fragments are not "
-              + "supported: module=" + module + ", moduleType=" + module.getModuleType());
+        IJ2EEModule j2eeModule = (IJ2EEModule) child.loadAdapter(IJ2EEModule.class, monitor);
+        IUtilityModule utilityModule =
+            (IUtilityModule) child.loadAdapter(IUtilityModule.class, monitor);
+        if (childDelegate == null || (j2eeModule == null && utilityModule == null)) {
+          logger.log(Level.WARNING, "child modules other than J2EE module or utility module are "
+              + "not supported: module=" + child + ", moduleType=" + child.getModuleType());
           continue;
         }
 
-        // per "isBinary()" Javadoc, "members()" should have a single resource.
-        IModuleResource zipResource = childDelegate.members()[0];
-        // destination (not an actual zip), e.g., "WEB-INF/lib/spring-web-4.3.6.RELEASE.jar"
-        IPath zip = new Path(delegate.getPath(child));
-        IPath zipParent = zip.removeLastSegments(1);
-
-        File javaIoFile = zipResource.getAdapter(File.class);
-        IFile iFile = zipResource.getAdapter(IFile.class);
-
-        if (javaIoFile != null) {
-          resources.add(new ModuleFile(javaIoFile, zipResource.getName(), zipParent));
-        } else if (iFile != null) {
-          resources.add(new ModuleFile(iFile, zipResource.getName(), zipParent));
+        // destination (relative to root), e.g., "WEB-INF/lib/spring-web-4.3.6.RELEASE.jar"
+        IPath destination = new Path(delegate.getPath(child));
+        String zipName = destination.lastSegment();  // to be created
+        IPath zipParent = destination.removeLastSegments(1);
+
+        boolean alreadyZip = j2eeModule != null && j2eeModule.isBinary()
+            || utilityModule != null && utilityModule.isBinary();
+        if (alreadyZip) {
+          // per "isBinary()" Javadoc, "members()" should have a single resource.
+          IModuleResource zipResource = childDelegate.members()[0];
+
+          File javaIoFile = zipResource.getAdapter(File.class);
```

Comment:
This seems weird. Why would either of these calls work once and not the next time? Or why would one fail and not the other?

Reply:
It's the Eclipse API that forces us to check both. `org.eclipse.wst.server.core.util.ModuleFile` with either the Java `File` or the Eclipse `IFile`:
```java
public class ModuleFile implements IModuleFile {
	private IFile file;
	private File file2;
        ...
	public ModuleFile(IFile file, String name, IPath path) { ...
          this.file = file;
          ... }

	public ModuleFile(File file, String name, IPath path) { ...
          this.file2 = file;
          ... }
```
We should be prepared to handle either use case, hence trying both.

Reply:
Where is IModuleResource.getAdapter documented? I can't find it.